File: src/main/java/cat/kittens/mods/controller/input/MappingActionExecutor.java
package cat.kittens.mods.controller.input;

import java.util.concurrent.atomic.AtomicLong;

public interface MappingActionExecutor {
    static MappingActionExecutor internal() {
        return (ctx) -> {
            throw new RuntimeException("This mapping is meant to be used internally by Mixin injections and cannot be executed directly.");
        };
    }

    static MappingActionExecutor withCooldown(long amountInMilliseconds, MappingActionExecutor callback) {
        AtomicLong lastExecution = new AtomicLong();
        return (ctx) -> {
            if (System.currentTimeMillis() > lastExecution.get() + amountInMilliseconds) {
                callback.perform(ctx);
                lastExecution.set(System.currentTimeMillis());
            }
        };
    }

    void perform(MappingExecutionContext ctx);
}

File: src/main/java/cat/kittens/mods/controller/input/MappingActionExecutor.kt
package cat.kittens.mods.controller.input

public fun interface MappingActionExecutor {
    public fun perform(ctx: MappingExecutionContext)

    public companion object {
        public fun internal(): MappingActionExecutor {
            return MappingActionExecutor {
                throw RuntimeException("This mapping is meant to be used internally by Mixin injections and cannot be executed directly.")
            }
        }

        public fun withCooldown(amountInMilliseconds: Long, callback: MappingActionExecutor): MappingActionExecutor {
            var lastExecution = -1L
            return MappingActionExecutor { ctx ->
                if (System.currentTimeMillis() > lastExecution + amountInMilliseconds) {
                    callback.perform(ctx)
                    lastExecution = System.currentTimeMillis()
                }
            }
        }
    }
}
